### ISSUE TITLE
Commands not reliably being sent to game

### ISSUE THREAD

**Issue by speedbird5** [WhiteMagic/JoystickGremlin#258]
Hi, I am using the temporary mode function assigned to a shift button to send different key stroke commands for other buttons depending on whether that shift button is being held.

However, in the game Falcon BMS (a flight simulator), the key strokes commands are being very unreliably and inconsistently sent to the game.  I have not tried different games.

﻿For example if you are using the temporary mode function by holding down a shift button, and let’s say you want **_ctrl + q_** sent to the game when you press the other desired button, half the time only **_q_** w﻿ould be﻿﻿ sent. 

Any idea why this might be happening ?

**Comment by WhiteMagic** [comment]
There are two scenarios I can think of. One is Gremlin is sending the keystrokes to quick and the other is that the way how Gremlin does remap to keyboard is backfiring.

The first problem you should be able to change the time between key sequences being sent in the Settings tab under "Default Macro Action Delay". Increasing this to 0.1 might work. The second possibility is that you are pressing and releasing the button very quicky and the way Gremlin currently handles the remap is that this causes it to release one button before the other one is pressed.

For example it could be that the game sees: Ctrl down, Ctrl up, q down, q up when you want it to see ctrl down, q down, q up, ctrl up. If that's the case then one way to work around this would be to use macros to send key strokes directly which would allow you to use them in exclusive mode which would prevent the "release" keys being sent before the "press" keys have been sent.

**Comment by speedbird5** [comment]
Thanks for the reply.

To be clear, I am not using macros but I am using the map to keyboard function to send ctrl + q

For the first problem:

Does map to keyboard actually record the speed and order I press and release the keys when setting the command ?

Are you saying the Default Macro Action Delay setting affects the map to keyboard function?

For the second problem using a Macro:

When I press and release a button, I need to be able to send, in the following order:
Ctrl down, q down, q up, ctrl up

Does this need to be programmed as two separate actions for button press and release ?

That is, ctrl down, q down for button press and q up, ctrl up for button release or can it all be done on button press ?

Do I use Keyboard or Joystick in the actions setting for the macro ?

For this macro, could you possibly post a screen shot of the macro using the example of ctrl + q

Thanks

**Comment by speedbird5** [comment]
Thanks for the reply. I am not using macros but simply the keyboard commands function. Does this actually record the speed and order I press the keys when setting the command ?

**Comment by speedbird5** [comment]
Sorry I accidentally closed this issue when I did not mean to.

**Comment by WhiteMagic** [comment]
Yeah, that setting affects the Map to Keyboard action, the reason for that is that behind the scenes Gremlin creates two macros that it runs when you press the button and the other one when you release it. How quickly you press buttons etc. isn't recorded by the action though which is why that setting might help.

I'll create an example macro setup tonight as it can be a bit confusing to setup if you don't know all the quirks Gremlin has due to how it uses things. Whether or not the macro would be two separate actions depends on exactly how you need to use it. If all you need is that ctrl + q is clicked and you don't require it to match the amount of time you hold the physical input down, then a single macro would be ideal. If you need the delay between pressing and releasing ctrl + q to match your physical button press then you would need two macros. Though I'll make examples of both use cases.

**Comment by speedbird5** [comment]
Thanks very much.  I am not sure which version will work but I will test both versions and report back.

I look forward to your examples.

**Comment by WhiteMagic** [comment]
Here are the two approaches I described above:

![bandicam 2019-11-09 15-29-07-856](https://user-images.githubusercontent.com/5612/68530153-15e15f00-0306-11ea-9d9b-22054a3bfd9e.png)
This one is a simple macro that will press and release the keys in the sequence lined out when you press the physical button. Nothing happens upon releasing the physical button, so effectively this is just sending that key combination the second you press the physical input.

![condition](https://user-images.githubusercontent.com/5612/68530190-707abb00-0306-11ea-9cc0-cea063648475.png)
The second one will perform a different action on pressing and releasing the physical input, which is achieved via the "Action Condition". This is effectively the same as what the "Map to Keyboard" action does behind the scenes. However, with the macro, you have the option to turn on the "Exclusive" flag which will prevent any other macro from running while that macro is being executed. This should help if for some reason your physical input actuation is really fast and Gremlin manages to release ctrl before q is being pressed.

I suspect that for your use case the first example should be sufficient, as I'm assuming Ctrl+q simply activates something and doesn't need holding for a certain amount of time.

**Comment by speedbird5** [comment]
Thanks very much. Yes I think the first case will be sufficient.

Do I need to increase the Default Macro Action Delay to 0.1 ?

Do I need the Exclusive flag for the first solution?

I will try them out and report back although it may be a couple of weeks until I get the chance.

**Comment by WhiteMagic** [comment]
Increasing the default delay should only be needed if there is a problem with the macro being executed properly. The exclusive flag would only really be useful if you are going to trigger the macro faster then macro's duration, though that seems unlikely.
